src/services/codefixes/convertToAsyncFunction.ts

```diff
@@ -0,0 +1,469 @@
+        const catchClause = createCatchClause(argName.identifier.text, createBlock(callbackBody));
+
+        const tryStatement = createTry(tryBlock, catchClause, /*finallyBlock*/ undefined);
+        return varDecl ? [varDecl, tryStatement] : [tryStatement];
+    }
+
+    function transformThen(node: CallExpression, transformer: Transformer, outermostParent: CallExpression, prevArgName?: SynthIdentifier): Statement[] {
+        const [res, rej] = node.arguments;
+
+        if (!res) {
+            return transformCallback(node.expression, transformer, outermostParent);
+        }
+
+        const argNameRes = getArgName(res, transformer);
+        const callbackBody = getCallbackBody(res, prevArgName, argNameRes, node, transformer);
+
+        if (rej) {
+            const argNameRej = getArgName(rej, transformer);
+
+            const tryBlock = createBlock(transformCallback(node.expression, transformer, node, argNameRes).concat(callbackBody));
+
+            const callbackBody2 = getCallbackBody(rej, prevArgName, argNameRej, node, transformer);
+            const catchClause = createCatchClause(argNameRej.identifier.text, createBlock(callbackBody2));
+
+            return [createTry(tryBlock, catchClause, /*finallyBlock*/ undefined) as Statement];
+        }
+        else {
+            return transformCallback(node.expression, transformer, node, argNameRes).concat(callbackBody);
+        }
+
+        return [];
+    }
+
+    function getFlagOfIdentifier(node: Identifier, constIdentifiers: Identifier[]): NodeFlags {
+        const inArr: boolean = constIdentifiers.some(elem => elem.text === node.text);
+        return inArr ? NodeFlags.Const : NodeFlags.Let;
+    }
+
+    function transformPromiseCall(node: Expression, transformer: Transformer, prevArgName?: SynthIdentifier): Statement[] {
+        const shouldReturn = transformer.setOfCallbacksToReturn.get(getNodeId(node).toString());
+        const hasPrevArgName = prevArgName && prevArgName.identifier.text.length > 0;
+        const originalNodeParent = node.original ? node.original.parent : node.parent;
+        if (hasPrevArgName && !shouldReturn && (!originalNodeParent || isPropertyAccessExpression(originalNodeParent))) {
+            return createVariableDeclarationOrAssignment(prevArgName!, createAwait(node), transformer).concat(); // hack to make the types match
+        }
+        else if (!hasPrevArgName && !shouldReturn && (!originalNodeParent || isPropertyAccessExpression(originalNodeParent))) {
+            return [createStatement(createAwait(node))];
+        }
+
+        return [createReturn(getSynthesizedDeepClone(node))];
+    }
+
+    function createVariableDeclarationOrAssignment(prevArgName: SynthIdentifier, rightHandSide: Expression, transformer: Transformer): NodeArray<Statement> {
+        if (prevArgName.numberOfUsesSynthesized < prevArgName.numberOfUsesOriginal) {
+            prevArgName.numberOfUsesSynthesized += 1;

```

Comment:
Personally, I would increment this regardless.  That would give you the option of asserting that the final case doesn't happen more than once.